addons/controllers/csi/vspherecsiconfig_utils.go

```diff
@@ -0,0 +1,259 @@
+
+	dvs.VSpherePVCSI.ClusterUID = string(cluster.UID) // only place to get UID
+
+	// default values from https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/vsphere-pv-csi/2.4.1/bundle/config/values.yaml
+	dvs.VSpherePVCSI.SupervisorMasterEndpointHostname = "supervisor.default.svc"
+	dvs.VSpherePVCSI.SupervisorMasterPort = 6443
+
+	// override derived & default values IF set in csi config
+	if config.Namespace != "" {
+		dvs.VSpherePVCSI.Namespace = config.Namespace
+	}
+
+	if config.ClusterName != "" {
+		dvs.VSpherePVCSI.ClusterName = config.ClusterName
+	}
+
+	if config.ClusterUID != "" {
+		dvs.VSpherePVCSI.ClusterUID = config.ClusterUID
+	}
+
+	if config.SupervisorMasterEndpointHostname != "" {
+		dvs.VSpherePVCSI.SupervisorMasterEndpointHostname = config.SupervisorMasterEndpointHostname
+	}
+
+	if config.SupervisorMasterPort != nil {
+		dvs.VSpherePVCSI.SupervisorMasterPort = *config.SupervisorMasterPort
+	}
+
+	return dvs, nil
+}
+
+func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValuesNonParavirtual(ctx context.Context,
+	vcsiConfig *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (*DataValues, error) {
+	dvs := &DataValues{}
+	config := vcsiConfig.Spec.VSphereCSI.NonParavirtualConfig
+
+	if config == nil {
+		return nil, errors.Errorf("NonParavirtualConfig missing in Spec.VSphereCSI")
+	}
+
+	// populate derived values
+	dvs.VSphereCSI = &DataValuesVSphereCSI{}
+
+	dvs.VSphereCSI.Namespace = tryParseClusterVariableString(ctx, cluster, NamespaceVarName)
+	dvs.VSphereCSI.ClusterName = tryParseClusterVariableString(ctx, cluster, ClusterNameVarName)
+	dvs.VSphereCSI.Server = tryParseClusterVariableString(ctx, cluster, VSphereServerVarName)
+	dvs.VSphereCSI.Datacenter = tryParseClusterVariableString(ctx, cluster, VSphereDatacenterVarName)
+	dvs.VSphereCSI.PublicNetwork = tryParseClusterVariableString(ctx, cluster, VSphereNetworkVarName) //TODO: IS 'VSPHERE_NETWORK' same as 'public network'?
+	dvs.VSphereCSI.Username = tryParseClusterVariableString(ctx, cluster, VSphereUsernameVarName)
+	dvs.VSphereCSI.Password = tryParseClusterVariableString(ctx, cluster, VSpherePasswordVarName)
+	dvs.VSphereCSI.Region = tryParseClusterVariableString(ctx, cluster, VSphereRegionVarName)
+	dvs.VSphereCSI.Zone = tryParseClusterVariableString(ctx, cluster, VSphereZoneVarName)
+	dvs.VSphereCSI.VSphereVersion = tryParseClusterVariableString(ctx, cluster, VSphereVersionVarName)
+	dvs.VSphereCSI.WindowsSupport = tryParseClusterVariableBool(ctx, cluster, IsWindowsWorkloadClusterVarName)
+	dvs.VSphereCSI.TLSThumbprint = tryParseClusterVariableString(ctx, cluster, VSphereTLSThumbprintVarName)
+
+	if cluster.Annotations != nil {
+		dvs.VSphereCSI.HttpProxy = cluster.Annotations[pkgtypes.HTTPProxyConfigAnnotation]
+		dvs.VSphereCSI.HttpsProxy = cluster.Annotations[pkgtypes.HTTPSProxyConfigAnnotation]
+		dvs.VSphereCSI.NoProxy = cluster.Annotations[pkgtypes.NoProxyConfigAnnotation]
```

Comment:
question: do we want to check if `cluster.Annotations[pkgtypes.HTTPProxyConfigAnnotation]` exist ? The case I could think of is when `cluster.Annotations != nil`, but there are no proxy related annotations, would it safe to set the proxy for vSphereCSI ?

Reply:
Not necessary. Golang will return an empty string from the map if the given key does not exist in it.